stdlib/public/TensorFlow/Ops.swift

```diff
@@ -1597,52 +1597,299 @@ public extension Tensor where Scalar : Numeric {
+  /// - Parameter upperBounds: The upper bounds at each dimension.
+  @inlinable
+  @differentiable(wrt: self)
+  func slice(lowerBounds: [Int32], upperBounds: [Int32]) -> Tensor {
+    /// TODO: Precondition `lowerBounds.count == upperBounds.count`,
+    /// preferably in graph.
+    let lowerBoundsTensor = Tensor<Int32>(lowerBounds)
+    return slice(
+      lowerBounds: lowerBoundsTensor,
+      sizes: Tensor<Int32>(upperBounds) - lowerBoundsTensor)
+  }
+
+  @inlinable
+  @differentiable(wrt: self, vjp: _vjpSlice)
+  func slice(lowerBounds: Tensor<Int32>, sizes: Tensor<Int32>) -> Tensor {
+    return Raw.slice(self, begin: lowerBounds, size: sizes)
+  }
+
   @inlinable
-  subscript(index: Int32) -> Tensor {
+  internal func _vjpSlice(
+    lowerBounds: Tensor<Int32>,
+    sizes: Tensor<Int32>
+  ) -> (Tensor, (Tensor) -> Tensor) {
+    let value = slice(lowerBounds: lowerBounds, sizes: sizes)
+    let afterPaddings = shapeTensor - value.shapeTensor - lowerBounds
+    return (value, { [after = afterPaddings] v in
+      let beforePaddings = lowerBounds.expandingShape(at: 1)
+      let afterPaddings = after.expandingShape(at: 1)
+      let paddings = Tensor<Int32>(
+        concatenating: [beforePaddings, afterPaddings], alongAxis: 1)
+      return Raw.pad(v, paddings: paddings)
+    })
+  }
+}
+
+public enum TensorRange : TensorRangeExpression {
+  case ellipsis
+  case newAxis
+  case squeezeAxis
+  case index(Int32)
+  case range(Range<Int32>, stride: Int32)
+  case closedRange(ClosedRange<Int32>, stride: Int32)
+  case partialRangeFrom(PartialRangeFrom<Int32>, stride: Int32)
+  case partialRangeUpTo(PartialRangeUpTo<Int32>, stride: Int32)
+  case partialRangeThrough(PartialRangeThrough<Int32>, stride: Int32)
+
+  public var tensorRange: TensorRange { return self }
+}
+
+extension TensorRange : Equatable {
+  public static func == (lhs: TensorRange, rhs: TensorRange) -> Bool {
+    switch (lhs, rhs) {
+    case (.ellipsis, .ellipsis),
+         (.newAxis, .newAxis),
+         (.squeezeAxis, .squeezeAxis):
+      return true
+    case (let .index(i1), let .index(i2)): return i1 == i2
+    case (let .range(r1, s1), let .range(r2, s2)): return r1 == r2 && s1 == s2
+    case (let .closedRange(r1, s1), let .closedRange(r2, s2)):
+      return r1 == r2 && s1 == s2
+    case (let .partialRangeFrom(r1, s1), let .partialRangeFrom(r2, s2)):
+      return r1.lowerBound == r2.lowerBound && s1 == s2
+    case (let .partialRangeUpTo(r1, s1), let .partialRangeUpTo(r2, s2)):
+      return r1.upperBound == r2.upperBound && s1 == s2
+    case (let .partialRangeThrough(r1, s1), let .partialRangeThrough(r2, s2)):
+      return r1.upperBound == r2.upperBound && s1 == s2
+    default: return false
+    }
+  }
+}
+
+public let ellipsis: TensorRange = .ellipsis
+public let newAxis: TensorRange = .newAxis
+public let squeezeAxis: TensorRange = .squeezeAxis
+
+public func strided(_ value: Range<Int32>, by stride: Int32) -> TensorRange {
```

Comment:
Hmm, I now think it's reasonable to just add an operator, since `strided(1..<3, by: 2)` looks really weird. Jokingly, I find `␣` to be a perfect operator for this.

How about using `..`? I think we need to add a new precedence group that is lower than `RangeFormationPrecedence` and higher than `CastingPrecedence`.

Reply:
Haha that's interesting. Yeah space could work even though people may miss it while skimming through code. And yes, I agree that I'd prefer we had an operator for this and was leaning towards `..` previously. I'll go ahead and add this for now.

Reply:
Thanks!